[clang][deps] Don't treat ObjC method args as module directives

Bigcheese opened the pull request:
`import:(type)name` is a method argument decl in ObjC, but the C++20 preprocessing rules say this is a preprocessing line.

Because the dependency directive scanner is not language dependent, this patch extends the C++20 rule to exclude `module :` (which is never a valid module decl anyway), and `import :` that is not followed by an identifier.

This is ok to do because in C++20 mode the compiler will later error on lines like this anyway, and the dependencies the scanner returns are still correct.

ChuanqiXu9 reviewed (approved):
LGTM as it looks like not affecting C++20 modules.

akyrtzi commented:
Could you test the change with a unit test in `DependencyDirectivesScannerTest.cpp` instead? It would be orders of magnitude more lightweight, execution wise, than a whole new lit test.

Bigcheese committed:
[clang][deps] Don't treat ObjC method args as module directives

`import:(type)name` is a method argument decl in ObjC, but the C++20
preprocessing rules say this is a preprocessing line.

Because the dependency directive scanner is not language dependent,
this patch extends the C++20 rule to exclude `module :` (which is
never a valid module decl anyway), and `import :` that is not followed
by an identifier.

This is ok to do because in C++20 mode the compiler will later error
on lines like this anyway, and the dependencies the scanner returns
are still correct.

Bigcheese commented:
Updated the test to be a unit test.

akyrtzi commented on clang/unittests/Lex/DependencyDirectivesScannerTest.cpp:
```diff
@@ -970,6 +970,26 @@ ort \
+      minimizeSourceToDependencyDirectives(Source, Out, Tokens, Directives));
+  // `module :` and `import :` not followed by an identifier are not treated as
+  // directive lines because they can be method argument decls.
+  EXPECT_EQ(Directives.size(), 2u); // 2 for the EOF tokens.
```
This would be a bit more clear if you checked against the `Out` variable, which contains a "pretty-printed" string of the directives that were found, something like this:

```
TEST(MinimizeSourceToDependencyDirectivesTest, ObjCMethodArgs) {
  SmallVector<char, 128> Out;

  StringRef Source = R"(
    @interface SomeObjcClass
      - (void)func:(int)otherData
              module:(int)module
              import:(int)import;
    @end
  )";

  ASSERT_FALSE(minimizeSourceToDependencyDirectives(Source, Out));
  // `module :` and `import :` not followed by an identifier are not treated as
  // directive lines because they can be method argument decls.
  EXPECT_STREQ("", Out.data());
}
```
